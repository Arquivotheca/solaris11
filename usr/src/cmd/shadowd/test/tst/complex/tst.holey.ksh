#!/bin/ksh -p
#
# CDDL HEADER START
#
# The contents of this file are subject to the terms of the
# Common Development and Distribution License (the "License").
# You may not use this file except in compliance with the License.
#
# You can obtain a copy of the license at usr/src/OPENSOLARIS.LICENSE
# or http://www.opensolaris.org/os/licensing.
# See the License for the specific language governing permissions
# and limitations under the License.
#
# When distributing Covered Code, include this CDDL HEADER in each
# file and include the License file at usr/src/OPENSOLARIS.LICENSE.
# If applicable, add the following below this CDDL HEADER, with the
# fields enclosed by brackets "[]" replaced with your own identifying
# information: Portions Copyright [yyyy] [name of copyright owner]
#
# CDDL HEADER END
#

#
# Copyright (c) 2009, 2011, Oracle and/or its affiliates. All rights reserved.
#

#
# Create a sparse file and verify that it is copied over correctly with holes
# intact.
#

. $ST_TOOLS/utility.ksh

function create_file
{
	typeset file=$1 size=$2
	shift 2

	mkfile -n $size $file || fail "failed to create $file"

	while [[ $# -gt 0 ]]; do
		typeset offset=$1 length=$2
		shift 2

		dd if=/dev/random of=$file bs=1 count=$length oseek=$offset \
		    conv=notrunc 2>/dev/null || fail "failed to write data"	
	done
}

function print_file
{
	typeset file=$1

	holey $file || fail "failed to get file contents"
}

tst_create_root_dataset -o recordsize=8k

#
# Create a file of the form:
#
#	DATA	8k
#	HOLE	16k
#	DATA	16k
#
create_file $TST_ROOT/a 40k 0 8k 24k 16k

#
# And another file that starts with a hole.  Ideally we want to end with a
# hole, but it's a bit of a pain because for reasons that aren't clear, the
# last block doesn't get turned into a hole on ZFS.
#
#	HOLE	16k
#	DATA	8k
#	HOLE	8k
#	DATA	64k
#
create_file $TST_ROOT/b 64k 16k 8k 32k 32k

print_file $TST_ROOT/a
echo
print_file $TST_ROOT/b
echo

tst_create_dataset -o recordsize=8k

print_file $TST_SROOT/a
echo
print_file $TST_SROOT/b

tst_destroy_dataset
tst_destroy_root_dataset
