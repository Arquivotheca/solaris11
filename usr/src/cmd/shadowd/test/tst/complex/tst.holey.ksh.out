DATA                0             2000
HOLE             2000             6000
DATA             6000             a000

HOLE                0             4000
DATA             4000             6000
HOLE             6000             8000
DATA             8000            10000

DATA                0             2000
HOLE             2000             6000
DATA             6000             a000

HOLE                0             4000
DATA             4000             6000
HOLE             6000             8000
DATA             8000            10000
